From e1e4ed776b857aca7e3c4bdeeccbebdeeb8ec0fa Mon Sep 17 00:00:00 2001
From: Mikhail Anikin <mikhail.anikin@solid-run.com>
Date: Sun, 27 Aug 2023 16:25:03 +0300
Subject: [PATCH 1/2] ravb: support multiple interfaces

Since ravb net driver uses obsolete mii bitbange driver,
the mii bus enumeration was broken and only last probed device worked
---
 drivers/net/ravb.c | 25 ++++++++++++++++++++++---
 include/miiphy.h   |  2 +-
 2 files changed, 23 insertions(+), 4 deletions(-)

diff --git a/drivers/net/ravb.c b/drivers/net/ravb.c
index 0a4cd2bf..ccb65961 100644
--- a/drivers/net/ravb.c
+++ b/drivers/net/ravb.c
@@ -538,6 +538,24 @@ static void ravb_stop(struct udevice *dev)
 	ravb_reset(dev);
 }
 
+// Glue to the legacy mii bitbang driver
+int ravb_bb_miiphy_read(struct mii_dev *bus, int addr, int devad, int reg)
+{
+	struct ravb_priv *priv = bus->priv;
+	bb_miiphy_buses[0].priv = priv;
+	sprintf(bb_miiphy_buses[0].name, bus->name);
+	return bb_miiphy_read(bus, addr, devad, reg);
+}
+
+int ravb_bb_miiphy_write(struct mii_dev *bus, int addr, int devad, int reg,
+						 u16 val)
+{
+	struct ravb_priv *priv = bus->priv;
+	bb_miiphy_buses[0].priv = priv;
+	sprintf(bb_miiphy_buses[0].name, bus->name);
+	return bb_miiphy_write(bus, addr, devad, reg, val);
+}
+
 static int ravb_probe(struct udevice *dev)
 {
 	struct eth_pdata *pdata = dev_get_plat(dev);
@@ -571,9 +589,10 @@ static int ravb_probe(struct udevice *dev)
 		goto err_mdio_alloc;
 	}
 
-	mdiodev->read = bb_miiphy_read;
-	mdiodev->write = bb_miiphy_write;
-	bb_miiphy_buses[0].priv = eth;
+	mdiodev->read = ravb_bb_miiphy_read;
+	mdiodev->write = ravb_bb_miiphy_write;
+	mdiodev->priv = eth;
+
 	snprintf(mdiodev->name, sizeof(mdiodev->name), dev->name);
 
 	ret = mdio_register(mdiodev);
diff --git a/include/miiphy.h b/include/miiphy.h
index 8b77bac0..7486fa2b 100644
--- a/include/miiphy.h
+++ b/include/miiphy.h
@@ -65,7 +65,7 @@ void mdio_list_devices(void);
 #define BB_MII_DEVNAME	"bb_miiphy"
 
 struct bb_miiphy_bus {
-	char name[16];
+	char name[32];
 	int (*init)(struct bb_miiphy_bus *bus);
 	int (*mdio_active)(struct bb_miiphy_bus *bus);
 	int (*mdio_tristate)(struct bb_miiphy_bus *bus);
-- 
2.41.0

